### ISSUE TITLE
Type error

### ISSUE THREAD

**Issue by kika** [AbdulRahmanAlHamali/flutter_typeahead#23]
I'm not sure if it's the Typeahead bug or Dart bug. 

Here's the repro case:
```dart
import 'dart:async';
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter_typeahead/flutter_typeahead.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'flutter_typeahead demo',
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return DefaultTabController(
      length: 1,
      child: Scaffold(
          appBar: AppBar(
            title: TabBar(tabs: [
              Tab(
                text: 'Example 1: Navigation',
              ),
            ]),
          ),
          body: TabBarView(children: [NavigationExample()])),
    );
  }
}

enum Skill {
    Math,
    English,
    Physics,
    Sports,
}

class NavigationExample extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: EdgeInsets.all(32.0),
      child: Column(
        children: <Widget>[
          SizedBox(
            height: 10.0,
          ),
          TypeAheadField<Skill>(
            textFieldConfiguration: TextFieldConfiguration(
              autofocus: true,
              decoration: InputDecoration(
                  border: OutlineInputBorder(),
                  hintText: 'Got skillz?'),
            ),
            suggestionsCallback: (pattern) => Skill.values,
            itemBuilder: (context, suggestion) =>
              ListTile(title: Text(suggestion.toString())),
            onSuggestionSelected: (suggestion) {
                print("Selected: ${suggestion.toString()}");
            },
          ),
        ],
      ),
    );
  }
}
```

If you run it as is it fails with exception
```
I/flutter (10939): The following assertion was thrown building
I/flutter (10939): _OverlayEntry-[LabeledGlobalKey<_OverlayEntryState>#c54a2](dirty, state: _OverlayEntryState#02217):
I/flutter (10939): type '(BuildContext, Skill) => ListTile' is not a subtype of type '(BuildContext, dynamic) =>
I/flutter (10939): Widget'
I/flutter (10939): Either the assertion indicates an error in the framework itself, or we should provide substantially
I/flutter (10939): more information in this error message to help you determine and fix the underlying cause.
I/flutter (10939): In either case, please report this assertion by filing a bug on GitHub:
I/flutter (10939):   https://github.com/flutter/flutter/issues/new?template=BUG.md
I/flutter (10939): When the exception was thrown, this was the stack:
I/flutter (10939): #0      _TypeAheadFieldState._initOverlayEntry.<anonymous closure> (package:flutter_typeahead/flutter_typeahead.dart:697:33)
I/flutter (10939): #1      _OverlayEntryState.build (package:flutter/src/widgets/overlay.dart:170:25)
```

If you remove `<Skill>` from the call to `TypeAheadField<Skill>(...)` it works fine.

```
Flutter 1.1.9 • channel dev • https://github.com/flutter/flutter.git
Framework • revision 1407091bfb (5 days ago) • 2019-01-08 20:40:19 -0800
Engine • revision e5ec3cf3ea
Tools • Dart 2.1.1 (build 2.1.1-dev.0.1 2cb346bd0c)
```

**Comment by Thetyne** [comment]
Reproduce with String type.

**Comment by Thetyne** [comment]
ok, i found the problem.
inside _initOverlayEntry method, the generic type T of "_SuggestionsList<T>" use the generic type of current class (_TypeAheadFieldState<T>).

When constuctor of _SuggestionsList is called, generic type T is unknow because it create inside a overlayEntry builder. the builder is executed outside class context.
So T begin dynamic inside _SuggestionsList.

Solution:
- Create _SuggestionsList  before call overlayEntry.

Code before:

```
Future<void> _initOverlayEntry() async {
    RenderBox renderBox = context.findRenderObject();
    var size = renderBox.size;

    this._suggestionsBoxController._overlayEntry =
        OverlayEntry(builder: (context) {
      return Positioned(
        width: size.width,
        child: CompositedTransformFollower(
          link: this._layerLink,
          showWhenUnlinked: false,
          offset:
              Offset(0.0, size.height + widget.suggestionsBoxVerticalOffset),
          child: _SuggestionsList<T>(
            suggestionsBoxController: _suggestionsBoxController,
            decoration: widget.suggestionsBoxDecoration,
            debounceDuration: widget.debounceDuration,
            controller: this._effectiveController,
            loadingBuilder: widget.loadingBuilder,
            noItemsFoundBuilder: widget.noItemsFoundBuilder,
            errorBuilder: widget.errorBuilder,
            transitionBuilder: widget.transitionBuilder,
            suggestionsCallback: widget.suggestionsCallback,
            animationDuration: widget.animationDuration,
            animationStart: widget.animationStart,
            getImmediateSuggestions: widget.getImmediateSuggestions,
            onSuggestionSelected: (T selection) {
              this._effectiveFocusNode.unfocus();
              widget.onSuggestionSelected(selection);
            },
            itemBuilder: widget.itemBuilder,
          ),
        ),
      );
    });
  }
```

Code after:
```
Future<void> _initOverlayEntry() async {
    RenderBox renderBox = context.findRenderObject();
    var size = renderBox.size;

    final _SuggestionsList<T> suggestionsList = _SuggestionsList<T>(
      suggestionsBoxController: _suggestionsBoxController,
      decoration: widget.suggestionsBoxDecoration,
      debounceDuration: widget.debounceDuration,
      controller: this._effectiveController,
      loadingBuilder: widget.loadingBuilder,
      noItemsFoundBuilder: widget.noItemsFoundBuilder,
      errorBuilder: widget.errorBuilder,
      transitionBuilder: widget.transitionBuilder,
      suggestionsCallback: widget.suggestionsCallback,
      animationDuration: widget.animationDuration,
      animationStart: widget.animationStart,
      getImmediateSuggestions: widget.getImmediateSuggestions,
      onSuggestionSelected: (T selection) {
        this._effectiveFocusNode.unfocus();
        widget.onSuggestionSelected(selection);
      },
      itemBuilder: widget.itemBuilder,
    );

    this._suggestionsBoxController._overlayEntry =
        OverlayEntry(builder: (context) {
      return Positioned(
        width: size.width,
        child: CompositedTransformFollower(
          link: this._layerLink,
          showWhenUnlinked: false,
          offset:
              Offset(0.0, size.height + widget.suggestionsBoxVerticalOffset),
          child: suggestionsList
        ),
      );
    });
  }
```

**Comment by sjmcdowall** [comment]
Want to submit a PR for that ??  Also .. the local variable in your example maybe should be _suggestionsList ...etc etc

**Comment by AbdulRahmanAlHamali** [comment]
Thank you for making the time to analyze the problem and find a solution. I am having problems with my laptop, so I will need a couple of days to implement the fix. Apologies for the delay.

**Comment by sjmcdowall** [comment]
I maybe able to do this -- but was wondering if the original author may want to submit a PR :)  .. Good sleuthing indeed

**Comment by kika** [comment]
Well, I've applied the patch manually (on my fork) and it didn't help. What confuses me is the wording of the error message:
`(BuildContext, String) => ListTile' is not a subtype of type '(BuildContext, dynamic) => Widget`

May be I'm missing something, but this statement is just not true. `String` is the subtype of `dynamic`, `ListTile` is the subtype of `Widget`. `BuildContext` is, well, `BuildContext`. So this error message is not true.

**Comment by Thetyne** [comment]
I sublit a PR today

**Comment by sjmcdowall** [comment]
@kika -- do you have strict type checking on (i.e. using the Google analyze.options?) -- if so there is something about dynamic being treated specially .. although I am not sure exactly how that implies .. but I've noticed similar things in other places in my code where the "linter" is complaining about something being not the same as something else, although its dynamic.  <shrug>

**Comment by kika** [comment]
Okay, I finally found it. Type errors in Dart are difficult to track.

**Comment by sjmcdowall** [comment]
@kika -- Ok... what was the solution? What did you find for others to see?

**Comment by kika** [comment]
@sjmcdowall You should see the PR, I'll try to link to it: #30

**Comment by sjmcdowall** [comment]
Ah great .. thank you

**Comment by sjmcdowall** [comment]
Fixed via #33
